clustpy/deep/dkm.py

```diff
@@ -0,0 +1,543 @@
+        # average losses
+        loss = (clean_loss + aug_loss) / 2
+        return loss
+
+    def _loss(self, batch: list, alpha: float, autoencoder: torch.nn.Module, cluster_loss_weight: float,
+              loss_fn: torch.nn.modules.loss._Loss, device: torch.device) -> torch.Tensor:
+        """
+        Calculate the complete DKM + Autoencoder loss.
+
+        Parameters
+        ----------
+        batch : list
+            the minibatch
+        alpha : float
+            the alpha value
+        autoencoder : torch.nn.Module
+            the autoencoder
+        cluster_loss_weight : float
+            weight of the clustering loss compared to the reconstruction loss
+        loss_fn : torch.nn.modules.loss._Loss
+            loss function for the reconstruction
+        device : torch.device
+            device to be trained on
+
+        Returns
+        -------
+        loss : torch.Tensor
+            the final DKM loss
+        """
+        # Calculate combined total loss
+        if self.augmentation_invariance:
+            # Calculate reconstruction loss
+            ae_loss, embedded, _ = autoencoder.loss([batch[0], batch[2]], loss_fn, device)
+            ae_loss_aug, embedded_aug, _ = autoencoder.loss([batch[0], batch[1]], loss_fn, device)
+            ae_loss = (ae_loss + ae_loss_aug) / 2
+            # Calculate clustering loss
+            cluster_loss = self.dkm_augmentation_invariance_loss(embedded, embedded_aug, alpha)
+        else:
+            # Calculate reconstruction loss
+            ae_loss, embedded, _ = autoencoder.loss(batch, loss_fn, device)
+            # Calculate clustering loss
+            cluster_loss = self.dkm_loss(embedded, alpha)
+        loss = ae_loss + cluster_loss * cluster_loss_weight
+        return loss
+
+    def fit(self, autoencoder: torch.nn.Module, trainloader: torch.utils.data.DataLoader, n_epochs: int,
+            device: torch.device, optimizer: torch.optim.Optimizer, loss_fn: torch.nn.modules.loss._Loss,
+            cluster_loss_weight: float) -> '_DKM_Module':
+        """
+        Trains the _DKM_Module in place.
+
+        Parameters
+        ----------
+        autoencoder : torch.nn.Module
+            the autoencoder
+        trainloader : torch.utils.data.DataLoader
+            dataloader to be used for training
+        n_epochs : int
+            number of epochs for the clustering procedure
+        device : torch.device
+            device to be trained on
+        optimizer : torch.optim.Optimizer
+            the optimizer for training
+        loss_fn : torch.nn.modules.loss._Loss
+            loss function for the reconstruction
+        cluster_loss_weight : float
+            weight of the clustering loss compared to the reconstruction loss
+
+        Returns
+        -------
+        self : _DKM_Module
+            this instance of the _DKM_Module
+        """
+        for alpha in self.alphas:
```

Comment:
Does the original implementation really have the alpha in the outer loop? If yes, then the documentation for n_epochs should change such that it is clear that it will be multiplied with len(self.alpha)

Reply:
Implementation follows the description given in the paper. 
Parameter documentation says: "number of epochs for each alpha value". Maybe not clear enough